File: src/services/junction/turn-state.test.mjs
import assert from 'node:assert/strict';
import test from 'node:test';
import {readFile} from 'node:fs/promises';
import path from 'node:path';
import {
  BLOCKER_ORDER, BLOCKER_TEXT, CUE_LABEL, FADER_START_CAPABILITY, FaderStart, SIGNAL_LABEL,
  deriveSignal, notOnAirReason, readyBlocker, tailLockReason, turnGuidance, turnSignal,
} from './turn-state.ts';

const nativeSource = () => readFile(path.resolve(import.meta.dirname, '../../../native/mixxx-engine-host/src/junction/turn_state.h'), 'utf8');
const roles = (local) => ({active: true, local, owner: 'a', next: 'b', outgoing: 'z'});
const allReady = () => ({compatible: true, programOpen: true, junctionRequired: true, junctionReceiving: true, junctionStable: true, latencyMeasured: true, pathLatencyMs: 80, budgetMs: 250, standbyStreamRequired: true, standbyStreamReceived: true, junctionUnity: true, localSilent: true});
const participant = (peerId, djName) => ({peerId, displayName: djName, djName});
const snapshot = (overrides = {}) => ({
  active: true, sessionId: 's', revision: 1, epoch: '2', lifecycle: 'live', localPeerId: 'b', hostPeerId: 'a', performerPeerId: 'a', operatorPeerId: 'a',
  handoffState: 'playing', participants: [participant('a', 'DJ A'), participant('b', 'DJ B'), participant('c', 'DJ C')],
  readiness: {ready: false, reasons: []}, program: {state: 'running'}, connection: {state: 'connected'}, ...overrides,
});
const turn = (overrides = {}) => ({signal: 'standby', nextPeerId: 'b', outgoingPeerId: '', blocker: {code: '', text: ''}, nextStatus: '', tailDecks: [], repeat: false, outOfQueue: [], incompatiblePeerIds: [], autoFailover: false, ...overrides});

test('signal light follows the roles exactly like native derive', () => {
  assert.equal(deriveSignal(roles('a'), false), 'onair');
  assert.equal(deriveSignal(roles('b'), false), 'standby');
  assert.equal(deriveSignal(roles('b'), true), 'ready');
  assert.equal(deriveSignal(roles('z'), true), 'outgoing');
  assert.equal(deriveSignal(roles('c'), true), 'off');
  assert.equal(deriveSignal({...roles('a'), active: false}, true), 'off');
  assert.deepEqual(Object.values(SIGNAL_LABEL), ['OFF', 'STANDBY', 'READY', 'ON AIR', 'OUTGOING']);
});

test('ready shows exactly one blocker in fix-first order', () => {
  assert.equal(readyBlocker(allReady()), '');
  assert.equal(readyBlocker({...allReady(), localSilent: false, junctionUnity: false}), 'junction_not_unity');
  assert.equal(readyBlocker({...allReady(), localSilent: false}), 'already_audible');
  assert.equal(readyBlocker({...allReady(), junctionReceiving: false, compatible: false}), 'update_required');
  assert.equal(readyBlocker({...allReady(), junctionReceiving: false}), 'waiting_junction');
  assert.equal(readyBlocker({...allReady(), junctionReceiving: false, junctionRequired: false}), '');
  assert.equal(readyBlocker({...allReady(), pathLatencyMs: 251}), 'latency_budget');
  assert.equal(readyBlocker({...allReady(), standbyStreamReceived: false}), 'host_waiting_stream');
  assert.equal(readyBlocker({...allReady(), standbyStreamReceived: false, standbyStreamRequired: false}), '');
});

test('native and TypeScript share the capability, blocker vocabulary, order and fader-start constants', async () => {
  const source = await nativeSource();
  assert.match(source, new RegExp(`kCapability="${FADER_START_CAPABILITY}"`));
  const codes = [...source.matchAll(/case Blocker::(\w+):return QStringLiteral\("([a-z_]+)"\);/g)].map((m) => m[2]);
  assert.deepEqual(codes, BLOCKER_ORDER, 'codes are listed in the same fix-first order');
  for (const code of BLOCKER_ORDER) assert(source.includes(`QStringLiteral("${BLOCKER_TEXT[code]}")`), `native text for ${code}`);
  const order = source.slice(source.indexOf('inline Blocker readyBlocker('), source.indexOf('inline QString blockerCode('));
  const returned = [...order.matchAll(/return Blocker::(\w+);/g)].map((m) => m[1]).filter((name) => name !== 'None');
  const enumOrder = [...source.matchAll(/case Blocker::(\w+):return QStringLiteral\("[a-z_]+"\);/g)].map((m) => m[1]);
  assert.deepEqual(returned, enumOrder, 'native readyBlocker checks in the published order');
  assert.match(source, /kThreshold=\.00316f/); assert.equal(FaderStart.threshold, 0.00316);
  assert.match(source, /kOnsetNanos=20000000LL/); assert.equal(FaderStart.onsetMs, 20);
  assert.match(source, /kArmNanos=300000000LL/); assert.equal(FaderStart.armMs, 300);
  assert.match(source, /残りの曲を送出中はマスター系を操作できません/); assert.match(source, /このデッキは残りの曲として送出中です。ループ以外は操作できません/);
});

test('fader start fires once on a silence to sound edge and never for a fader already up', () => {
  const fader = new FaderStart();
  for (let t = 0; t <= 1000; t += 5) assert.equal(fader.observe(0.5, false, t), false);
  assert.equal(fader.armed, false);
  for (let t = 1005; t < 1300; t += 5) assert.equal(fader.observe(0, false, t), false);
  assert.equal(fader.observe(0, false, 1305), false); assert.equal(fader.armed, true);
  assert.equal(fader.observe(0.01, false, 1310), false); assert.equal(fader.observe(0.01, false, 1325), false);
  assert.equal(fader.observe(0.01, false, 1330), true); assert.equal(fader.observe(0.01, false, 1335), false);
  const j = new FaderStart(); j.observe(0, false, 0); j.observe(0, false, 300);
  assert.equal(j.observe(0, true, 301), false); assert.equal(j.observe(0, true, 321), true);
});

test('tail lock matches native: loops only on tail decks, everything free elsewhere', () => {
  for (const op of ['deck.loop.set', 'deck.loop.enable', 'mixer.channel.pfl']) assert.equal(tailLockReason(op, {deck: 'A'}, ['A']), '');
  for (const op of ['deck.play', 'deck.seek', 'deck.hotcue.jump', 'deck.tempo.set', 'deck.sync.set', 'deck.load', 'mixer.channel.gain', 'mixer.filter.set']) assert.notEqual(tailLockReason(op, {deck: 'A'}, ['A']), '');
  for (const op of ['deck.play', 'deck.load', 'mixer.channel.gain']) assert.equal(tailLockReason(op, {deck: 'B'}, ['A']), '');
  assert.notEqual(tailLockReason('mixer.crossfader', {}, ['A']), '');
  assert.notEqual(tailLockReason('deck.sync.set', {deck: 'B', leader: 'A'}, ['A']), '');
});

test('guidance is one DJ-facing line without system vocabulary', () => {
  const standby = snapshot({turn: turn()});
  assert.equal(turnSignal(standby), 'standby');
  assert.equal(turnGuidance(standby), 'STANDBY — DJ Aの音を受信中。曲を用意してCUEで合わせてください');
  assert.equal(turnGuidance(snapshot({turn: turn({signal: 'ready'})})), 'READY — フェーダーを上げるとON AIR');
  assert.equal(turnGuidance(snapshot({localPeerId: 'a', turn: turn({signal: 'onair', outgoingPeerId: 'c'})})), 'ON AIR — DJ Cの曲が残っています。Jを下げると交代完了');
  assert.equal(turnGuidance(snapshot({localPeerId: 'c', operatorPeerId: 'b', performerPeerId: 'b', turn: turn({signal: 'outgoing', outgoingPeerId: 'c'})})), 'OUTGOING — 残りの曲をDJ Bが下げるまで送っています（ループのみ操作可）');
  assert.equal(turnGuidance(snapshot({turn: turn({blocker: {code: 'already_audible', text: ''}})})), 'STANDBY — 一度フェーダーを下げてください');
  assert.equal(turnGuidance(snapshot({operatorPeerId: '', performerPeerId: '', turn: turn()})), 'STANDBY — 最初のDJです。曲を用意してください');
  assert.equal(notOnAirReason(snapshot({turn: turn({blocker: {code: 'latency_budget', text: ''}})})), 'まだ本番に出ていません：回線の遅延が大きく、会場の音に間に合いません');
  assert.equal(notOnAirReason(snapshot({turn: turn({signal: 'ready'})})), '');
  for (const line of [turnGuidance(standby), ...Object.values(BLOCKER_TEXT), ...Object.values(CUE_LABEL)]) {
    assert.doesNotMatch(line, /操作権|送出中|direct-stream|feedbackBlocked|epoch/);
  }
});

File: native/mixxx-engine-host/tests/junction/turn_state_test.cpp
#include "harness.h"
#include "junction/turn_state.h"
using namespace junction::turn;
namespace {
Roles roles(const QString& local){Roles r;r.active=true;r.local=local;r.owner="a";r.next="b";r.outgoing="z";return r;}
ReadyInputs allReady(){ReadyInputs in;in.junctionReceiving=true;in.junctionStable=true;in.latencyMeasured=true;in.pathLatencyMs=80;in.budgetMs=250;in.standbyStreamRequired=true;in.standbyStreamReceived=true;in.junctionUnity=true;in.localSilent=true;return in;}
QJsonObject unity(){return {{"available",true},{"volume",1.0},{"eqLow",1.0},{"eqMid",1.0},{"eqHigh",1.0},{"orientation",1}};}
}
JTEST("turn","signal light follows the existing roles"){
 CHECK(derive(roles("a"),false)==Signal::OnAir);
 CHECK(derive(roles("b"),false)==Signal::Standby);
 CHECK(derive(roles("b"),true)==Signal::Ready);
 CHECK(derive(roles("z"),true)==Signal::Outgoing);
 CHECK(derive(roles("c"),true)==Signal::Off);
 auto inactive=roles("a");inactive.active=false;CHECK(derive(inactive,true)==Signal::Off);
 CHECK_EQ(signalName(Signal::Ready),QString("ready"));CHECK_EQ(signalName(Signal::OnAir),QString("onair"));
}
JTEST("turn","ready shows exactly one blocker in fix-first order"){
 CHECK(readyBlocker(allReady())==Blocker::None);
 auto in=allReady();in.localSilent=false;in.junctionUnity=false;CHECK(readyBlocker(in)==Blocker::JunctionNotUnity);
 in.junctionUnity=true;CHECK(readyBlocker(in)==Blocker::AlreadyAudible);CHECK_EQ(blockerText(readyBlocker(in)),QString("一度フェーダーを下げてください"));
 in=allReady();in.junctionReceiving=false;in.compatible=false;CHECK(readyBlocker(in)==Blocker::UpdateRequired);
 in.compatible=true;CHECK(readyBlocker(in)==Blocker::WaitingJunction);
 in.junctionRequired=false;CHECK(readyBlocker(in)==Blocker::None);
 in=allReady();in.pathLatencyMs=251;CHECK(readyBlocker(in)==Blocker::LatencyBudget);
 in=allReady();in.standbyStreamReceived=false;CHECK(readyBlocker(in)==Blocker::HostWaitingStream);in.standbyStreamRequired=false;CHECK(readyBlocker(in)==Blocker::None);
 in=allReady();in.programOpen=false;in.latencyMeasured=false;CHECK(readyBlocker(in)==Blocker::ProgramClosed);
 for(auto b:{Blocker::UpdateRequired,Blocker::ProgramClosed,Blocker::WaitingJunction,Blocker::JunctionUnstable,Blocker::LatencyUnknown,Blocker::LatencyBudget,Blocker::HostWaitingStream,Blocker::JunctionNotUnity,Blocker::AlreadyAudible}){CHECK(!blockerCode(b).isEmpty());CHECK(!blockerText(b).isEmpty());}
}
JTEST("turn","fader start fires once on a silence to sound edge"){
 CHECK_NEAR(FaderStart::kThreshold,.00316,.00001);CHECK_EQ(FaderStart::kOnsetNanos,20000000LL);CHECK_EQ(FaderStart::kArmNanos,300000000LL);
 FaderStart f;const qint64 ms=1000000;
 // Arriving with a fader up never fires, however long it stays up.
 for(qint64 t=0;t<=1000;t+=5)CHECK(!f.observe(.5f,false,t*ms));
 CHECK(!f.armed());
 for(qint64 t=1005;t<1300;t+=5)CHECK(!f.observe(0,false,t*ms));
 CHECK(!f.armed());CHECK(!f.observe(0,false,1305*ms));CHECK(f.armed());
 CHECK(!f.observe(.01f,false,1310*ms));CHECK(!f.observe(.01f,false,1325*ms));
 CHECK(f.observe(.01f,false,1330*ms));
 CHECK(!f.observe(.01f,false,1335*ms));CHECK(!f.armed());
}
JTEST("turn","a short click does not fire and a J move counts as sound"){
 FaderStart f;const qint64 ms=1000000;
 CHECK(!f.observe(0,false,0));CHECK(!f.observe(0,false,400*ms));CHECK(f.armed());
 CHECK(!f.observe(.2f,false,405*ms));CHECK(!f.observe(0,false,410*ms));CHECK(!f.observe(.2f,false,415*ms));CHECK(!f.observe(.2f,false,430*ms));
 CHECK(f.observe(.2f,false,435*ms));
 FaderStart j;CHECK(!j.observe(0,false,0));CHECK(!j.observe(0,false,300*ms));CHECK(!j.observe(0,true,301*ms));CHECK(j.observe(0,true,321*ms));
 CHECK(!j.observe(FaderStart::kThreshold*.9f,false,400*ms));
}
JTEST("turn","J unity accepts only unity flat THRU"){
 CHECK(junctionUnity(unity()));
 auto c=unity();c["volume"]=.9;CHECK(!junctionUnity(c));
 c=unity();c["eqLow"]=0.0;CHECK(!junctionUnity(c));
 c=unity();c["orientation"]=0;CHECK(!junctionUnity(c));
 c=unity();c["available"]=false;CHECK(!junctionUnity(c));
}
JTEST("turn","latency budget covers the relay and the standby stream"){
 CHECK_NEAR(LatencyBudget::budgetMs(24000),250,.001);
 CHECK_NEAR(LatencyBudget::pathMs(100,40,false),100+20+LatencyBudget::kMarginMs,.001);
 CHECK_NEAR(LatencyBudget::pathMs(100,40,true),100,.001);
}
JTEST("turn","tail decks accept loops only and the rest of the mixer stays free"){
 const QSet<int> tail{0};
 const auto deck=[](const char* d){return QJsonObject{{"deck",d}};};
 for(const char* op:{"deck.loop.set","deck.loop.enable","deck.loop.beat","mixer.channel.pfl"})CHECK(TailLock::check(op,deck("A"),tail).isEmpty());
 for(const char* op:{"deck.play","deck.seek","deck.hotcue.jump","deck.scratch","deck.pitchbend","deck.tempo.set","deck.key.shift","deck.sync.set","deck.load","deck.unload","deck.beatjump","mixer.channel.gain","mixer.channel.eq","mixer.eq.set","mixer.filter.set","mixer.fx.set","mixer.colorfx.set","mixer.channel.orientation","mixer.trim.set"})
  CHECK(!TailLock::check(op,deck("A"),tail).isEmpty());
 for(const char* op:{"deck.play","deck.load","mixer.channel.gain","deck.sync.set"})CHECK(TailLock::check(op,deck("B"),tail).isEmpty());
 for(const char* op:{"mixer.crossfader","mixer.master.gain","mixer.beatfx.set"})CHECK(!TailLock::check(op,{},tail).isEmpty());
 CHECK(!TailLock::check("deck.sync.set",QJsonObject{{"deck","B"},{"leader","A"}},tail).isEmpty());
 CHECK(TailLock::check("sampler.play",{},tail).isEmpty());
}
JTEST("turn","tail ends only after the tail stops and stays silent"){
 TailEnded t;const qint64 s=1000000000LL;
 CHECK(!t.observe(true,0,0));CHECK(!t.observe(false,.1f,1*s));CHECK(!t.observe(false,0,2*s));CHECK(!t.observe(false,0,4*s));
 CHECK(!t.observe(false,.5f,4*s+1));CHECK(!t.observe(false,0,5*s));CHECK(t.observe(false,0,8*s));
}
JTEST("turn","timetable picks the next eligible DJ and repeat keeps B2B looping"){
 const QStringList order{"a","b","c","d"};
 const auto all=[](const QString&){return true;};
 CHECK_EQ(nextInOrder(order,"a","",{},all),QString("b"));
 CHECK_EQ(nextInOrder(order,"a","b",{},all),QString("c"));
 CHECK_EQ(nextInOrder(order,"a","",{"b"},all),QString("c"));
 CHECK_EQ(nextInOrder(order,"a","",{},[](const QString& id){return id=="d";}),QString("d"));
 CHECK_EQ(nextInOrder(QStringList{"a"},"a","",{},all),QString());
 CHECK(afterTurn(order,"b","a",{})==(QStringList{"b","c","d","a"}));
 CHECK(afterTurn(QStringList{"a","b","c","d"},"b","a",{"d"})==(QStringList{"b","c","a","d"}));
 CHECK(afterTurn(QStringList{"a","b","c","d"},"b","a",{"a","d"})==(QStringList{"b","c","d","a"}));
 CHECK(afterTurn(QStringList{"a","b"},"b","a",{})==(QStringList{"b","a"}));
 CHECK(joined(QStringList{"a","b","c","d"},"a","d",{"c","d"})==(QStringList{"a","b","d","c"}));
 CHECK(joined(QStringList{"a","c"},"a","c",{"c"})==(QStringList{"a","c"}));
 CHECK(skipped(QStringList{"a","b","c","d"},"a","b",{"d"})==(QStringList{"a","c","b","d"}));
 CHECK(skipped(QStringList{"a","b"},"a","a",{})==(QStringList{"a","b"}));
}
